stellargraph/mapper/link_mappers.py

```diff
@@ -163,6 +162,120 @@ def on_epoch_end(self):
+                   (The graph nodes must have a "feature" attribute that is used as input to the GraphSAGE model.)
+
+        sampler: An UnsupervisedSampler instance that indicates how the neighborhood of graph are sampled (uniform random walks, biased random walks etc. Currently only Uniform Random walks are enabled).
+                This class is responsible for calling the right type of walk generator and then return batch_size of sample source and target pairs  from those walks.
+                These (target, context) pairs are to be used to train or inference, and the embeddings
+                calculated for the links created by a binary operator applied to the target and context nodes,
+                are passed to the downstream task of link prediction or link attribute inference.
+                The target and context nodes of the links are used as head nodes for which subgraphs are sampled.
+                The subgraphs are sampled from all nodes.
+
+  """
+
+    def __init__(self, generator, walker):
+
+        self.generator = generator  # graphlinkgenerator instance
+
+        self.ids = (
+            []
+        )  # Since this is an instance of on demand sampling, at the initialization we don't have the pregenerated head samples.
+
+        if isinstance(walker, UnsupervisedSampler):
+
+            self.walker = walker
+
+            self.data_size = (
+                len(self.walker.nodes)
+                * self.walker.length
+                * self.walker.number_of_walks
+            )  # an estimate of the  upper bound on how many samples are generated in each epoch
+
+            print(
+                "Running GraphSAGELinkGenerator with an estimated {} batches generated on the fly per epoch.".format(
+                    round(self.data_size / self.generator.batch_size)
+                )
+            )
+
+            self._gen = self.walker.generator(
+                self.generator.batch_size
+            )  # the generator method from the sampler with the batch-size from the link generator method
+        else:
+            raise TypeError(
+                "({}) UnsupervisedSampler is required.".format(type(self).__name__)
+            )
+
+    def __getitem__(self, batch_num):
+        """
+        Generate one batch of data.
+
+        Args:
+            batch_num<int>: number of a batch
+
+        Returns:
+            batch_feats<list>: Node features for nodes and neighbours sampled from a
+                batch of the supplied IDs
+            batch_targets<list>: Targets/labels for the batch.
+
+        """
+
+        if batch_num >= self.__len__():
+            raise IndexError(
+                "Mapper: batch_num larger than number of esstaimted  batches for this epoch."
+            )
+        # print("Fetching {} batch {} [{}]".format(self.name, batch_num, start_idx))
+
+        # Get head nodes and labels
+        head_ids, batch_targets = next(self._gen)
+        self.ids = list(head_ids)
```

Comment:
Do we need to set `self.ids` for each batch?

Reply:
Removed the self.ids variable. 
This was the class variable that was being passed to the method that infer the head node types. Now, the head node types are inferred only when the first batch that is generated. I pass the head_ids of the first batch for head node type inference.